File: js/dailies/06_aug_18/quickSort.js
const randomArr = require('../../helpers/randomArr');

const partition = (arr, begIdx, endIdx) => {
	const randIdx = Math.floor(Math.random() * (endIdx - begIdx)) + begIdx;
	swap(arr, begIdx, randIdx);

	let pivotIdx = begIdx;
	const pivotVal = arr[endIdx];

	for (let i = begIdx; i < endIdx; i++) {
		if (arr[i] < pivotVal) {
			swap(arr, i, pivotIdx++);
		}
	}

	swap(arr, pivotIdx, endIdx);

	return pivotIdx;
};

const quickSort = (arr, begIdx = 0, endIdx = arr.length - 1) => {
	if (begIdx < endIdx) {
		const pivot = partition(arr, begIdx, endIdx);
		quickSort(arr, begIdx, pivot - 1);
		quickSort(arr, pivot + 1, endIdx);
	}
};

const swap = (arr, x, y) => {
	const temp = arr[x];
	arr[x] = arr[y];
	arr[y] = temp;
};

const myArr = randomArr(true, 10);
console.log(myArr);
quickSort(myArr);
console.log(myArr);

module.exports = quickSort;

File: python/dailies/4_nov_22/quick_sort.py
import math
import random

from typing import List, Optional
from helpers.testing.list_sort_tester import ListSortTester


def _swap(lst: List[int], x: int, y: int) -> None:
    temp = lst[x]
    lst[x] = lst[y]
    lst[y] = temp


def _partition(
    lst: List[int], beg_idx: int, end_idx: int
) -> int:
    rand_idx = math.floor(
        (end_idx - beg_idx) * random.random()
    ) + beg_idx

    _swap(lst, rand_idx, end_idx)

    pivot_idx = beg_idx
    pivot_val = lst[end_idx]

    for i in range(beg_idx, end_idx):
        if lst[i] < pivot_val:
            _swap(lst, i, pivot_idx)

            pivot_idx += 1

    _swap(lst, pivot_idx, end_idx)

    return pivot_idx


def quick_sort(
        lst: List[int],
        beg_idx: Optional[int] = None,
        end_idx: Optional[int] = None,
) -> List[int]:
    if beg_idx is None:
        beg_idx = 0
        end_idx = len(lst) - 1

    if beg_idx < end_idx:
        pivot = _partition(lst, beg_idx, end_idx)

        lst = quick_sort(lst, beg_idx, pivot - 1)
        lst = quick_sort(lst, pivot + 1, end_idx)

    return lst


ListSortTester(quick_sort)
